### PR TITLE
html entities encode/decode

### PR DESCRIPTION
**Pull request by neraliu**
this PR includes the following
- build the trie data structure for named char reference look up
- provide the public api of encode/decode for html entities encode/decode
- use the latest html5 named character reference list for html decode.
https://html.spec.whatwg.org/multipage/entities.json

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: src/html-entities.js [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        var saveFile = path+"/data/entities.json.save",
+            jsonFile = path+"/data/entities.json";
+
+        if (fs.existsSync(saveFile)) {
+            this.loadNamedCharReferenceTrie(saveFile);
+        } else if (fs.existsSync(jsonFile)) {
+            var d = fs.readFileSync(jsonFile, "utf8"),
+                o = JSON.parse(d);
+            this.buildNamedCharReferenceTrie(o);
+        }
```
Comment (neraliu) [inline src/html-entities.js]: will throw error if fail to load or file not exist.

File: src/html-entities.js [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+            return '&'+s;
+        } else {
+            var radix = (isHex !== 'x' && isHex !== 'X')? 10:16;
+            var i = parseInt(p1, radix);
+            // 0-55295, 57344-65535, 65536-1114111
+            if ((i>=0 && i<=0xD7FF) || (i>=0xE000 && i<=0xFFFF) || (i>=0x10000 && i<=0x10FFFF)) {
+                return htmlDecoder.fromCodePoint(i);
+            } else {
+               if (isNaN(i)) {
+                   return '&#'+s; // not html entities
```
Comment (neraliu) [inline src/html-entities.js]: should we combine it and return the case '&#'+s only?

File: src/html-entities.js [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+* @description
+* HTML decode the character
+*/
+HTMLEntities.prototype.reCharReferenceDecode = /&(#?)([xX]?)0*([a-zA-Z0-9]+)(;)?/g;
+HTMLEntities.prototype.fromCodePoint = String.fromCodePoint;
+HTMLEntities.prototype.decode = function(str) {
+    /* decode all numeric dec/hex character reference */
+    var htmlDecoder = this;
+
+    str = str.replace(this.reCharReferenceDecode, function(m, isNamed, isHex, p1, p2) {
```
Comment (adon-at-work) [inline src/html-entities.js]: `&000gt;` will be converted as `>`

File: src/html-entities.js (hunk 2 above)
Comment (adon-at-work) [inline src/html-entities.js]: chrome does that

File: src/html-entities.js (hunk 3 above)
Comment (adon-at-work) [inline src/html-entities.js]: I think we can capture those that we strictly need by:

```javascript
HTMLEntities.prototype.reCharReferenceDecode = /&([a-z];?)|&#0*((x)[a-f0-9]|[0-9]);?/ig;
```

```javascript
HTMLEntities.prototype.decode = function(str) {
var self = this;

return str.replace(this.reCharReferenceDecode, function(m, named, number) {
    if (named) {
        named = self._findString(named);
        return named ? named.characters : m;
    } else {
        var i = parseInt(number[0] <= '9' ? number : '0' + number);  // parseInt('0xA0') is equiv to parseInt('A0', 16)
        return ((i>=0 && i<=0xD7FF) || (i>=0xE000 && i<=0xFFFF) || (i>=0x10000 && i<=0x10FFFF)) ? htmlDecoder.fromCodePoint(i) : '\uFFFD';
    }
});

};
```

File: src/html-entities.js [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+         Adonis Fung <adon@yahoo-inc.com>
+*/
+/*jshint -W030 */
+/*jshint eqnull:true */
+(function () {
+"use strict";
+
+var fs = require('fs');
+/* polyfill for codePointAt, fromCodePoint */
+require('./polyfill.js');
```
Comment (adon-at-work) [inline src/html-entities.js]: I don't think you need this, as NodeJS already have everything in polyfill.js

Comment (adon-at-work) [inline src/html-entities.js]: check out the grunt file and you'll see polyfill.js will be included only for building dist/...js

Comment (neraliu) [inline src/html-entities.js]: there is no need in the client side. what do u mean of the NodeJS (server side)?

Comment (adon-at-work) [inline src/html-entities.js]: you code can run using NodeJS without `require('./polyfill.js')`, right?

File: src/html-entities.js [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+/**
+* @function HTMLEntities#_findStringWithCodePoint
+*
+* @description
+*/
+HTMLEntities.prototype._findStringWithCodePoint = function(trie, str, pos) {
+    /* skip the '&' */
+    if (str[pos] === '&') 
+        return this._findStringWithCodePoint(trie, str, pos+1);
```
Comment (adon-at-work) [inline src/html-entities.js]: pos++ will do.

File: src/html-entities.js (hunk 4 above)
Comment (neraliu) [inline src/html-entities.js]: we can't

**Comment by neraliu** [comment]
it worths effort to decouple it as a standalone package.

**Pull request closed without merging**